enhancements/network/persistent-ips-sdn-secondary-networks.md

```diff
@@ -0,0 +1,1012 @@
+    podAllocator->>podAllocator:annotatedPod(IPs)
+```
+
+##### Delete VM - IP pool flow
+```mermaid
+sequenceDiagram
+  actor user
+  participant gc
+  participant kubevirt
+  box Blue ovnkube-control-plane
+    participant ipamLeaseAllocator
+    participant ipAllocator
+  end
+
+  user->>kubevirt: delete VM
+  kubevirt-->>user: OK
+  gc->>ipamLeaseAllocator: Delete(ipamLease)
+  ipamLeaseAllocator->>ipAllocator:ReleaseIPs(ipamLease.IPs)
+  ipAllocator-->>ipamLeaseAllocator:Result
+  ipamLeaseAllocator-->user:Result
+```
+
+#### How to integrate the new CRD with the OVN-K IP pool on IPAMLease delete
+
+When a persistentIP allocation is removed from the datastore (via Kubernetes GC
+), we must return said IP address to the pool, by invoking `ReleaseIPs` for
+every IP in it.
+
+### Kubernetes Garbage Collection extended info
+
+This feature is relying on Kubernetes garbage collection to delete the
+persistent IP allocations once the Virtual Machine object is deleted.
+
+But, we actually need to assure the persistent IP allocation stays around as
+long as the running Virtual Machine (i.e. the process) is running - thus, we
+also need to ensure the pod where the VM runs was deleted. If it weren't, we
+will end up in a situation where the IP would be returned to the pool earlier
+than it should, which would lead to the same IP in two workloads at the same
+time: one of them shutting down, the other already readily processing traffic.
+
+#### KubeVirt finalizers
+
+The VM has a finalizer that prevents it from being deleted while the Virtual
+Machine Instance (VMI, represents a running VM) is around; the VMI - in turn -
+has a finalizer that prevents it from being deleted until the pod where the VM
+runs is deleted.
+
+Different
+[DeletionPropagation](https://github.com/kubernetes/apimachinery/blob/3e2600dc79feea6cdc8a9224bc8a6a7fcfee1466/pkg/apis/meta/v1/types.go#L526)
+values will impact how Kubernetes will garbage collect the `IPAMLease`s when
+deleting the VM object from the datastore. The default deletion propagation is
+`background`.
+
+#### Deleting the VM with background `DeletionPropagation`
+
+The following sequence diagram describes how Kubernetes GC works for the
+KubeVirt entities (VM / VMI / pod):
+
+```mermaid
+sequenceDiagram
+  actor User
+  participant persistentIP
+  participant GC
+  participant VM
+  participant VMI
+  participant launcherPod
+
+  User->>VM: Delete VM
+  VM->>VMI: Delete VMI
+  activate VM
+  VMI->>launcherPod: Delete launcherPod
+  activate VMI
+  launcherPod-->>VMI: launcherPod Deleted
+  Deactivate VMI
+  note over VMI: clear VMI finalizer
+  VMI-->>VM: VMI Deleted
+  Deactivate VM
+  note over VM: clear VM finalizer
+  VM-->>User: VM Deleted
+  VM->>GC: VM deleted
+  GC->>persistentIP: Delete persistentIP
+  persistentIP-->>GC: persistentIP Deleted
+```
+
+The two finalizers mentioned above ensure the VM object is only removed from
+the datastore after the VM has actually stopped. Since the `IPAMLease`
```

Comment:
With background delete, the children are GCed only when the owner is gone.
Once the user delete the VM, the vm-controller will delete the VMI if exists, and once the VMI is gone (happens after pod is gone with the same logic), it will remove the VM finalizer [1].
Until the VM finalizer is there, the VM won't be deleted, once the finalizer is removed, k8s will be allowed to clean the VM, and then the GC can kick in and clean the owned resources, such as IPAM Lease.

[1] https://github.com/kubevirt/kubevirt/blob/4accdebe16489462937e2d52f8e233736a8acfa4/pkg/virt-controller/watch/vm.go#L2594